File: 202.happy-number.java
/*
 * @lc app=leetcode id=202 lang=java
 *
 * [202] Happy Number
 */
class Solution {

    public boolean isHappy(int n) {
        Set<Integer> history = new HashSet<>();
        while (history.add(n)) {
            int sum = 0;
            while (n > 0) {
                sum += sq(n % 10);
                n /= 10;
            }
            if (sum == 1) return true;
            n = sum;
        }
        return false;
    }

    private int sq(int a) {
        return a * a;
    }
}

File: go/202.happy-number.go
/*
 * @lc app=leetcode id=202 lang=golang
 *
 * [202] Happy Number
 */

// @lc code=start

type void struct{}

var member void

func isHappy(n int) bool {
	set := make(map[int]void)
	for {
		sum := 0
		for n != 0 {
			r := n % 10
			sum += r * r
			n /= 10
		}
		if sum == 1 {
			return true
		}
		if _, b := set[sum]; b {
			return false
		} else {
			set[sum] = member
		}
		n = sum
	}
}

// @lc code=end

